File: src/path.zig
// path: pure path-resolution helpers.
//
// The kernel keeps its VFS resolver absolute-only (vfs.resolve in
// src/vfs.zig). Per-task `cwd` (task_layout.TaskStruct.cwd) lives one
// abstraction layer above — sys_chdir stores into it, and sys_openFile
// / execveKernel join relative paths against it at the syscall
// boundary before handing the absolute result to vfs.resolve.
//
// joinResolve is the single non-recursive `.` / `..` collapse used
// for both store (sys_chdir) and resolve (open/execve). Pure: no
// allocator, no externs, no kernel imports — exercised in isolation
// by the host suite (see the tests at the bottom of this file).
//
// Sibling-import only — no named module needed for path.zig itself.
// build.zig wires the comptime-pure source for both the kernel and
// host builds; consumers that name this file as a named import
// (`sys.zig`, `execve.zig`) reach it through the `path` module wired
// in build.zig.

const std = @import("std");

// Maximum nesting depth the component stack can hold. 64 components is
// well above the working set (paths like `/bin/fsh` or
// `/etc/fshrc` have one or two components); a pathological caller
// gets a clean null instead of writing past the stack.
const MAX_DEPTH: usize = 64;

// Working buffer for the (cwd + "/" + rel) composition before the
// single-pass collapse. 512 bytes covers a 256-byte cwd plus a
// 255-byte relative tail with room for the joiner slash, which is the
// ceiling sys_chdir's user-side copy enforces.
const WORK_MAX: usize = 512;

// joinResolve: normalise `rel` against `cwd` into `out`. Returns a
// slice of `out` containing the absolute path with `.` / `..` /
// duplicate-slash segments collapsed. `out` must be large enough to
// hold the final result; returns null on:
//   * `rel` already absolute (leading '/') AND longer than WORK_MAX
//   * cwd + '/' + rel longer than WORK_MAX
//   * resolved length wouldn't fit in `out`
//   * deeper than MAX_DEPTH components after collapse
//
// `cwd` is treated as an absolute base; an empty `cwd` resolves to
// "/" (defensive — callers should pass at least "/"). A leading '/'
// in `rel` bypasses cwd entirely (still gets collapsed). Trailing
// slashes are dropped. The empty resolved path is normalised to "/"
// so callers can blindly hand the result to vfs.resolve.
pub fn joinResolve(cwd: []const u8, rel: []const u8, out: []u8) ?[]const u8 {
    var work: [WORK_MAX]u8 = undefined;
    var work_len: usize = 0;

    if (rel.len > 0 and rel[0] == '/') {
        if (rel.len > WORK_MAX) return null;
        @memcpy(work[0..rel.len], rel);
        work_len = rel.len;
    } else {
        // Anchor on cwd; "" → "/" (defensive). Always emit a slash
        // separator before splicing in `rel` — the collapse below
        // treats repeated slashes as a single boundary.
        if (cwd.len == 0) {
            if (work_len + 1 > WORK_MAX) return null;
            work[work_len] = '/';
            work_len += 1;
        } else {
            if (cwd.len > WORK_MAX) return null;
            @memcpy(work[0..cwd.len], cwd);
            work_len = cwd.len;
        }
        if (work_len == 0 or work[work_len - 1] != '/') {
            if (work_len + 1 > WORK_MAX) return null;
            work[work_len] = '/';
            work_len += 1;
        }
        if (work_len + rel.len > WORK_MAX) return null;
        @memcpy(work[work_len..][0..rel.len], rel);
        work_len += rel.len;
    }

    // Component stack: each entry is the byte offset in `out` where
    // the leading '/' of that component begins. Popping `..` restores
    // out_len to the stored offset, which discards the just-pushed
    // path segment in a single assignment.
    var stack: [MAX_DEPTH]usize = undefined;
    var depth: usize = 0;
    var out_len: usize = 0;

    var i: usize = 0;
    while (i < work_len) {
        // Skip any run of slashes (folds "//" + leading "/").
        while (i < work_len and work[i] == '/') i += 1;
        if (i >= work_len) break;
        var j = i;
        while (j < work_len and work[j] != '/') j += 1;
        const comp = work[i..j];
        i = j;

        if (comp.len == 1 and comp[0] == '.') {
            // Skip a "." segment.
        } else if (comp.len == 2 and comp[0] == '.' and comp[1] == '.') {
            if (depth > 0) {
                depth -= 1;
                out_len = stack[depth];
            }
            // ".." past root is a no-op (stays at "/").
        } else {
            if (depth >= MAX_DEPTH) return null;
            if (out_len + 1 + comp.len > out.len) return null;
            stack[depth] = out_len;
            depth += 1;
            out[out_len] = '/';
            out_len += 1;
            @memcpy(out[out_len..][0..comp.len], comp);
            out_len += comp.len;
        }
    }

    if (out_len == 0) {
        if (out.len < 1) return null;
        out[0] = '/';
        out_len = 1;
    }
    return out[0..out_len];
}

// ---- Host tests ----

const testing = std.testing;

test "joinResolve: relative against root" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/", "bin/fsh", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/bin/fsh", r);
}

test "joinResolve: relative against non-root cwd" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/etc", "fshrc", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/etc/fshrc", r);
}

test "joinResolve: absolute rel bypasses cwd" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/etc", "/bin/fsh", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/bin/fsh", r);
}

test "joinResolve: dot segments are dropped" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/usr", "./local/./bin", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/usr/local/bin", r);
}

test "joinResolve: parent collapses one component" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/usr/local/bin", "../lib", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/usr/local/lib", r);
}

test "joinResolve: mid-path .. collapses correctly" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/", "a/./b/../c", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/a/c", r);
}

test "joinResolve: .. past root stays at root" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/", "../../foo", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/foo", r);
}

test "joinResolve: bare .. from root is root" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/", "..", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/", r);
}

test "joinResolve: double slashes fold" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/foo", "//bar//baz", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/bar/baz", r);
}

test "joinResolve: empty rel resolves to cwd" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/etc", "", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/etc", r);
}

test "joinResolve: trailing slash dropped" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/", "etc/", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/etc", r);
}

test "joinResolve: dot-only rel resolves to cwd" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/var/log", ".", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/var/log", r);
}

test "joinResolve: collapse to root when popping everything" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("/a/b", "../..", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/", r);
}

test "joinResolve: out-buffer overflow returns null" {
    var tiny: [4]u8 = undefined;
    try testing.expectEqual(@as(?[]const u8, null), joinResolve("/", "abcdefg", &tiny));
}

test "joinResolve: oversize composition returns null" {
    var buf: [4096]u8 = undefined;
    const long_rel: [WORK_MAX + 1]u8 = .{'x'} ** (WORK_MAX + 1);
    try testing.expectEqual(@as(?[]const u8, null), joinResolve("/", long_rel[0..], &buf));
}

test "joinResolve: empty cwd resolves under root" {
    var buf: [128]u8 = undefined;
    const r = joinResolve("", "foo", &buf) orelse return error.UnexpectedNull;
    try testing.expectEqualStrings("/foo", r);
}

File: crates/kernel/src/path.rs
//! Allocation-free absolute path resolution.

const MAX_DEPTH: usize = 64;
const WORK_MAX: usize = 512;

/// Normalize `rel` against `cwd` into `out`.
///
/// Duplicate slashes and `.` components are removed, `..` pops one component
/// without crossing root, and an absolute `rel` bypasses `cwd`.
pub fn join_resolve<'a>(cwd: &[u8], rel: &[u8], out: &'a mut [u8]) -> Option<&'a mut [u8]> {
    let mut work = [0u8; WORK_MAX];
    let work_len;

    if rel.first() == Some(&b'/') {
        if rel.len() > WORK_MAX {
            return None;
        }
        work[..rel.len()].copy_from_slice(rel);
        work_len = rel.len();
    } else {
        let mut len = if cwd.is_empty() {
            work[0] = b'/';
            1
        } else {
            if cwd.len() > WORK_MAX {
                return None;
            }
            work[..cwd.len()].copy_from_slice(cwd);
            cwd.len()
        };
        if len == 0 || work[len - 1] != b'/' {
            if len == WORK_MAX {
                return None;
            }
            work[len] = b'/';
            len += 1;
        }
        let end = len.checked_add(rel.len())?;
        if end > WORK_MAX {
            return None;
        }
        work[len..end].copy_from_slice(rel);
        work_len = end;
    }

    let mut stack = [0usize; MAX_DEPTH];
    let mut depth = 0;
    let mut out_len = 0;
    let mut i = 0;

    while i < work_len {
        while i < work_len && work[i] == b'/' {
            i += 1;
        }
        if i >= work_len {
            break;
        }
        let mut j = i;
        while j < work_len && work[j] != b'/' {
            j += 1;
        }
        let component = &work[i..j];
        i = j;

        if component == b"." {
            continue;
        }
        if component == b".." {
            if depth > 0 {
                depth -= 1;
                out_len = stack[depth];
            }
            continue;
        }
        if depth >= MAX_DEPTH {
            return None;
        }
        let end = out_len.checked_add(1)?.checked_add(component.len())?;
        if end > out.len() {
            return None;
        }
        stack[depth] = out_len;
        depth += 1;
        out[out_len] = b'/';
        out[out_len + 1..end].copy_from_slice(component);
        out_len = end;
    }

    if out_len == 0 {
        if out.is_empty() {
            return None;
        }
        out[0] = b'/';
        out_len = 1;
    }
    Some(&mut out[..out_len])
}

#[cfg(test)]
mod tests {
    use super::{join_resolve, WORK_MAX};

    fn resolves(cwd: &[u8], rel: &[u8]) -> [u8; 128] {
        let mut out = [0u8; 128];
        let len = join_resolve(cwd, rel, &mut out).unwrap().len();
        out[len] = 0;
        out
    }

    fn value(buf: &[u8]) -> &[u8] {
        &buf[..buf.iter().position(|b| *b == 0).unwrap()]
    }

    #[test]
    fn relative_against_root() {
        assert_eq!(value(&resolves(b"/", b"bin/fsh")), b"/bin/fsh");
    }
    #[test]
    fn relative_against_non_root_cwd() {
        assert_eq!(value(&resolves(b"/etc", b"fshrc")), b"/etc/fshrc");
    }
    #[test]
    fn absolute_rel_bypasses_cwd() {
        assert_eq!(value(&resolves(b"/etc", b"/bin/fsh")), b"/bin/fsh");
    }
    #[test]
    fn dot_segments_are_dropped() {
        assert_eq!(
            value(&resolves(b"/usr", b"./local/./bin")),
            b"/usr/local/bin"
        );
    }
    #[test]
    fn parent_collapses_one_component() {
        assert_eq!(
            value(&resolves(b"/usr/local/bin", b"../lib")),
            b"/usr/local/lib"
        );
    }
    #[test]
    fn mid_path_parent_collapses_correctly() {
        assert_eq!(value(&resolves(b"/", b"a/./b/../c")), b"/a/c");
    }
    #[test]
    fn parent_past_root_stays_at_root() {
        assert_eq!(value(&resolves(b"/", b"../../foo")), b"/foo");
    }
    #[test]
    fn bare_parent_from_root_is_root() {
        assert_eq!(value(&resolves(b"/", b"..")), b"/");
    }
    #[test]
    fn double_slashes_fold() {
        assert_eq!(value(&resolves(b"/foo", b"//bar//baz")), b"/bar/baz");
    }
    #[test]
    fn empty_rel_resolves_to_cwd() {
        assert_eq!(value(&resolves(b"/etc", b"")), b"/etc");
    }
    #[test]
    fn trailing_slash_is_dropped() {
        assert_eq!(value(&resolves(b"/", b"etc/")), b"/etc");
    }
    #[test]
    fn dot_only_rel_resolves_to_cwd() {
        assert_eq!(value(&resolves(b"/var/log", b".")), b"/var/log");
    }
    #[test]
    fn popping_everything_collapses_to_root() {
        assert_eq!(value(&resolves(b"/a/b", b"../..")), b"/");
    }
    #[test]
    fn out_buffer_overflow_returns_none() {
        let mut tiny = [0u8; 4];
        assert!(join_resolve(b"/", b"abcdefg", &mut tiny).is_none());
    }
    #[test]
    fn oversize_composition_returns_none() {
        let mut out = [0u8; 4096];
        let long_rel = [b'x'; WORK_MAX + 1];
        assert!(join_resolve(b"/", &long_rel, &mut out).is_none());
    }
    #[test]
    fn empty_cwd_resolves_under_root() {
        assert_eq!(value(&resolves(b"", b"foo")), b"/foo");
    }
}
